File: core/src/main/java/com/terralite/core/registry/RegistryManager.java
package com.terralite.core.registry;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

public final class RegistryManager {
    private final Map<RegistryKey<?>, MutableRegistry<?>> registries = new LinkedHashMap<>();
    private boolean frozen;

    public <T> MutableRegistry<T> create(RegistryKey<T> key) {
        Objects.requireNonNull(key, "key");

        if (frozen) {
            throw new IllegalStateException("Registry manager is frozen");
        }

        if (registries.containsKey(key)) {
            throw new IllegalArgumentException("Duplicate registry: " + key.id());
        }

        MutableRegistry<T> registry = new SimpleMutableRegistry<>(key);
        registries.put(key, registry);
        return registry;
    }

    public <T> MutableRegistry<T> requireMutable(RegistryKey<T> key) {
        MutableRegistry<?> registry = registries.get(key);
        if (registry == null) {
            throw new IllegalArgumentException("Missing registry: " + key.id());
        }
        return cast(key, registry);
    }

    public GameData freeze() {
        frozen = true;

        Map<RegistryKey<?>, FrozenRegistry<?>> frozenRegistries = new LinkedHashMap<>();
        for (Map.Entry<RegistryKey<?>, MutableRegistry<?>> entry : registries.entrySet()) {
            frozenRegistries.put(entry.getKey(), entry.getValue().freeze());
        }

        return new GameData(frozenRegistries);
    }

    @SuppressWarnings("unchecked")
    private static <T> MutableRegistry<T> cast(RegistryKey<T> key, MutableRegistry<?> registry) {
        if (!registry.key().type().equals(key.type())) {
            throw new IllegalStateException("Registry type mismatch for " + key.id());
        }
        return (MutableRegistry<T>) registry;
    }
}

File: core/src/commonMain/kotlin/com/terralite/core/registry/RegistryManager.kt
package com.terralite.core.registry

/**
 * Accumulates [MutableRegistry] instances and seals them all into a [GameData] snapshot.
 *
 * Registries must be created before [freeze] is called; afterwards no new registries
 * or entries can be added.
 */
class RegistryManager {

    private val registries: LinkedHashMap<RegistryKey<*>, MutableRegistry<*>> = LinkedHashMap()
    private var frozen = false

    fun <T> create(key: RegistryKey<T>): MutableRegistry<T> {
        check(!frozen) { "Registry manager is frozen" }
        require(!registries.containsKey(key)) { "Duplicate registry: ${key.id}" }
        val registry = SimpleMutableRegistry(key)
        registries[key] = registry
        return registry
    }

    fun <T> requireMutable(key: RegistryKey<T>): MutableRegistry<T> {
        val registry = registries[key]
            ?: throw IllegalArgumentException("Missing registry: ${key.id}")
        return cast(key, registry)
    }

    fun freeze(): GameData {
        frozen = true
        val frozen = LinkedHashMap<RegistryKey<*>, FrozenRegistry<*>>()
        for ((key, registry) in registries) {
            frozen[key] = registry.freeze()
        }
        return GameData(frozen)
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> cast(key: RegistryKey<T>, registry: MutableRegistry<*>): MutableRegistry<T> {
        return registry as MutableRegistry<T>
    }
}
